lib/class-wp-theme-json.php

```diff
@@ -0,0 +1,967 @@
+	 *   '--wp--nested-property--sub-property': 'value'
+	 * }
+	 *
+	 * @param array  $tree Input tree to process.
+	 * @param string $prefix Prefix to prepend to each variable. '' by default.
+	 * @param string $token Token to use between levels. '--' by default.
+	 *
+	 * @return array The flattened tree.
+	 */
+	private function flatten_tree( $tree, $prefix = '', $token = '--' ) {
+		$result = array();
+		foreach ( $tree as $property => $value ) {
+			$new_key = $prefix . str_replace(
+				'/',
+				'-',
+				strtolower( preg_replace( '/(?<!^)[A-Z]/', '-$0', $property ) ) // CamelCase to kebab-case.
+			);
+
+			if ( is_array( $value ) ) {
+				$new_prefix = $new_key . $token;
+				$result     = array_merge(
+					$result,
+					$this->flatten_tree( $value, $new_prefix, $token )
+				);
+			} else {
+				$result[ $new_key ] = $value;
+			}
+		}
+		return $result;
+	}
+
+	/**
+	 * Utility to extract a path from an array.
+	 *
+	 * @param array $array    Array we want to retrieve data from.
+	 * @param array $path     Array containing the path to retrieve.
+	 * @param array $default  Value to return if $array doesn't contain the $path.
+	 *
+	 * @return array Data at the given $path or $default.
+	 */
+	private function get_from_path( $array, $path, $default = array() ) {
+		if ( ! is_array( $array ) || ! is_array( $path ) ) {
+			return $default;
+		}
+
+		$path_length = count( $path );
+		for ( $i = 0; $i < $path_length; ++$i ) {
+			if ( ! isset( $array[ $path[ $i ] ] ) ) {
+				return $default;
+			}
+			$array = $array[ $path[ $i ] ];
+		}
+		return $array;
+	}
+
+	/**
+	 * Returns the style property for the given path.
+	 *
+	 * It also converts CSS Custom Property stored as
+	 * "var:preset|color|secondary" to the form
+	 * "--wp--preset--color--secondary".
+	 *
+	 * @param array $styles Styles subtree.
+	 * @param array $path Which property to process.
+	 *
+	 * @return string Style property value.
+	 */
+	private function get_property_value( $styles, $path ) {
+		$value = $this->get_from_path( $styles, $path, '' );
+
+		if ( '' === $value ) {
+			return $value;
+		}
+
+		$prefix     = 'var:';
+		$prefix_len = strlen( $prefix );
+		$token_in   = '|';
+		$token_out  = '--';
+		if ( 0 === strncmp( $value, $prefix, $prefix_len ) ) {
+			$unwrapped_name = str_replace(
+				$token_in,
+				$token_out,
+				substr( $value, $prefix_len )
+			);
+			$value          = "var(--wp--$unwrapped_name)";
+		}
+
+		return $value;
+	}
+
+	/**
+	 * Given a context, it extracts the style properties
+	 * and adds them to the $declarations array following the format:
+	 *
+	 * ```php
+	 * array(
+	 *   'name'  => 'property_name',
+	 *   'value' => 'property_value,
+	 * )
+	 * ```
+	 *
+	 * Note that this modifies the $declarations in place.
+	 *
+	 * @param array $declarations Holds the existing declarations.
+	 * @param array $context Input context to process.
+	 */
+	private function compute_style_properties( &$declarations, $context ) {
```

Comment:
Should this function be static?

Reply:
I don't have a strong opinion on this, I'm fine marking them as static. I'm curious why it'd be preferred, though!

Reply:
https://github.com/WordPress/gutenberg/pull/26803/commits/718e367302cbb4d2c4b39eb7afb7b53e3c076a22

Reply:
I think it is a good practice to have all the functions that don't depend on the instance of a class marked as static.